core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java

```diff
@@ -0,0 +1,496 @@
+    ListenableFuture<Certificate> ephemeralCertificateFuture =
+        executor.submit(this::fetchEphemeralCertificate);
+
+    ListenableFuture<SSLContext> sslContextFuture =
+        whenComplete(
+            () ->
+                createSslContext(
+                    Futures.getDone(metadataFuture), Futures.getDone(ephemeralCertificateFuture)),
+            metadataFuture,
+            ephemeralCertificateFuture);
+
+    ListenableFuture<InstanceData> refreshFuture =
+        whenComplete(
+            () ->
+                new InstanceData(
+                    Futures.getDone(metadataFuture), Futures.getDone(sslContextFuture)),
+            metadataFuture,
+            sslContextFuture);
+
+    return refreshFuture;
+  }
+
+  /**
+   * Creates a new SSLContext based on the provided parameters. This SSLContext will be used to
+   * provide new SSLSockets that are authorized to connect to a Cloud SQL instance.
+   */
+  private SSLContext createSslContext(Metadata metadata, Certificate ephemeralCertificate) {
+    SSLContext sslContext;
+    try {
+      KeyStore authKeyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      authKeyStore.load(null, null);
+      KeyStore.PrivateKeyEntry privateKey =
+          new PrivateKeyEntry(keyPair.getPrivate(), new Certificate[] {ephemeralCertificate});
+      authKeyStore.setEntry("ephemeral", privateKey, null);
+      KeyManagerFactory kmf =
+          KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm());
+      kmf.init(authKeyStore, null);
+
+      KeyStore trustedKeyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+      trustedKeyStore.load(null, null);
+      trustedKeyStore.setCertificateEntry("instance", metadata.getInstanceCaCertificate());
+      TrustManagerFactory tmf = TrustManagerFactory.getInstance("X.509");
+      tmf.init(trustedKeyStore);
+
+      sslContext = SSLContext.getInstance("TLSv1.2");
+      sslContext.init(kmf.getKeyManagers(), tmf.getTrustManagers(), new SecureRandom());
+
+    } catch (GeneralSecurityException | IOException ex) {
+      // TODO(kvg): Verify this should be unchecked
+      throw new RuntimeException(
+          String.format(
+              "[%s] Unable to create a SSLContext for the Cloud SQL instance.", connectionName),
+          ex);
+    }
+    return sslContext;
+  }
+
+  /** Fetches the latest version of the instance's metadata using the Cloud SQL Admin API. */
+  private Metadata fetchMetadata() {
+    DatabaseInstance instanceMetadata;
+    try {
+      instanceMetadata = apiClient.instances().get(projectId, instanceId).execute();
+    } catch (IOException ex) {
+      throw addExceptionContext(
+          ex,
+          String.format("[%s] Failed to update metadata for Cloud SQL instance.", connectionName));
+    }
+
+    // Validate the instance will support the authenticated connection.
+    if (!instanceMetadata.getRegion().equals(regionId)) {
+      throw new IllegalArgumentException(
```

Comment:
So in this case, this implies that the argument to the _socketFactory_ is bad - the user has provided an incorrect connection name.

I really have no idea - this is left over behavior from the previous version. It might be possible to have multiple instances with the same instance name but different regions, but I'm not positive.